### PR TITLE
Test Differ#pick_encoding; move to EncodedString

### PR DESCRIPTION
**Pull request by bf4**
Followup to https://github.com/rspec/rspec-support/pull/151

especially see https://github.com/rspec/rspec-support/pull/134#issuecomment-68984440

### CODE REVIEW

File: lib/rspec/support/differ.rb [hunk 1]
```diff
@@ -7,6 +7,7 @@ module RSpec
   module Support
     # rubocop:disable ClassLength
     class Differ
+      LINE_BREAK = "\n"
```
Comment (JonRowe) [inline lib/rspec/support/differ.rb]: What's the reasoning behind this?

Comment (bf4) [inline lib/rspec/support/differ.rb]: Well, as I commented in https://github.com/bf4/rspec-support/commit/ce333af71a110fe5f080b3b0e8cc37c8cb05db6a I was thinking of using `LINE_BREAK` as a way to get the `__ENCODING__` of strings generated in `differ.rb`, that also happened to be a nice cleanup of string generation. Turns out it didn't work like I thought it did, so, the only reason to keep it now is cleanup, if you like. I'm going to try again with a less aggressive strategy.

File: lib/rspec/support/differ.rb [hunk 2]
```diff
@@ -95,18 +96,18 @@ def diffably_stringify(array)
           if Array === entry
             entry.inspect
           else
-            entry.to_s.gsub("\n", "\\n")
+            entry.to_s.gsub(LINE_BREAK, Regexp.escape(LINE_BREAK))
```
Comment (JonRowe) [inline lib/rspec/support/differ.rb]: So the pre-escaped `"\\n"` still seems like a better idea than running an escape each time this is used.

File: lib/rspec/support/differ.rb [hunk 3]
```diff
@@ -182,9 +183,9 @@ def object_to_string(object)
             pp_value = PP.singleline_pp(object[key], "")
 
             "#{pp_key} => #{pp_value},"
-          end.join("\n")
+          end.join(LINE_BREAK)
         when String
-          object =~ /\n/ ? object : object.inspect
+          object =~ /#{LINE_BREAK}/o ? object : object.inspect
```
Comment (JonRowe) [inline lib/rspec/support/differ.rb]: Same here, leaving the primitive seems better than interpolating the constant each time.

**Comment by bf4** [comment]
I thinking of simplifying this PR to 

1. Testing `Differ#pick_encoding`
2. Moving `pick_encoding` into `EncodedString`

Then, in another PR, per any discussion in this one, change the tests in the `differ_spec.rb` to use an encoding-agnostic `be_identical_string` matcher, such as in this [changset](https://github.com/bf4/rspec-support/commit/6ab674a9fbc9102daf1aaa411413fb1ffa4d46e8#diff-298595d31b0fe8bb190912673bab6659R28-R35), or just use object equality, for now.

**Comment by bf4** [comment]
Since I'm going to revise this PR a bit, I thought I might as well push the ArgumentError on EncodingString#split fix for a preview of what it is, though it would go into a subsequent PR.

File: lib/rspec/support/spec/encoded_string_matcher.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  chain :with_encoding do |enc|
+    @expected_encoding = enc
+  end
+
+  # HACK: To disable the encoding check
+  chain :without_encoding do
+    @without_encoding = :without_encoding
+  end
```
Comment (bf4) [inline lib/rspec/support/spec/encoded_string_matcher.rb]: re: the hack here and below, I'm wondering if I should change the name of the matcher to `expect_identical_encoded_string` and `expect_identical_string` and see if there are still any cases for chaining

**Comment by bf4** [comment]
I want to keep this moving, so I've reduced this PR to just the Differ/EncodedString#pick_encoding

**Comment by JonRowe** [comment]
Thanks @bf4

**Comment by JonRowe** [comment]
(Note @myronmarston I haven't added a change log because this is pretty much just an internal refactoring)

**Pull request merged**